File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/C++/huffman_coding.cpp
#include<iostream>
#include<string>
#include<queue>
#include<unordered_map>
using namespace std; 

class HuffmanNode{
public:  char ch; int freq ; HuffmanNode *left,*right;
HuffmanNode(char c,int f): ch(c),freq(f),left(NULL),right(NULL){}
}; 

class Compare{
public: 
    bool operator()(HuffmanNode* a,HuffmanNode* b){
        return a->freq > b->freq; }
};

class HuffmanCoding{
    HuffmanNode* root; unordered_map<char,string>codes;
public: 
    HuffmanCoding(){root=nullptr;}

    void build_tree(const unordered_map<char,int>& freq_map){
        priority_queue<HuffmanNode*,vector<HuffmanNode*>,Compare>min_heap;
        for(auto& pair: freq_map)min_heap.push(new HuffmanNode(pair.first,pair.second));
        while(min_heap.size()>1){
            HuffmanNode* left = min_heap.top(); min_heap.pop();
            HuffmanNode* right = min_heap.top(); min_heap.pop();
            HuffmanNode* merged = new HuffmanNode('\0',left->freq+right->freq);
            merged->left=left; merged->right=right;
            min_heap.push(merged);
        }
        root = min_heap.top();
    }

    void assign_codes(){
        codes.clear();
        assign_codes_helper(root,"");
    }
    void assign_codes_helper(HuffmanNode*node,string code){
        if(!node)return; 
        if(node->ch != '\0'){
            codes[node->ch]=code; return;
        }
        assign_codes_helper(node->left,code+'0');
        assign_codes_helper(node->right,code+'1');
    }

    string encode(const string& text){
        string encoded; 
        for(char ch: text)encoded+=codes[ch];
        return encoded; 
    }

    string decode(const string& encoded){
        string decoded; 
        HuffmanNode* current = root; 
        for(char bit: encoded){
            if(bit=='0')current=current->left;
            else current=current->right;
            if(current->ch != '\0'){
                decoded+=current->ch; 
                current=root; 
            }
        }
        return decoded; 
    }

    void print_codes(){
        for(auto& pair: codes)cout<<pair.first<<": "<<pair.second<<endl; 
    }
};
int main() 
{ 
     unordered_map<char,int> freq_map = {{'A',3},{'B',4},{'C',6},{'D',2},{'E',2}};
     HuffmanCoding huffman; 
     huffman.build_tree(freq_map);
     huffman.assign_codes();
     cout<<"Huffman Codes: "<<endl;
     huffman.print_codes();
     string text = "AABCCDEEDDEC";
     string encoded = huffman.encode(text);  cout<<"Encoded Text: "<<encoded<<endl; 
     string decoded = huffman.decode(encoded); cout<<"Decoded Text: "<<decoded<<endl;
}

File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/Python/huffman_coding.py
import heapq 
from typing import Dict, Optional, Tuple, List

class HuffmanNode:
    def __init__(self,char: Optional[str], freq: int):
        self.char = char
        self.freq = freq
        self.left: Optional['HuffmanNode'] = None
        self.right: Optional['HuffmanNode'] = None
    def __lt__(self, other: 'HuffmanNode') -> bool:
        return self.freq < other.freq
    
class HuffmanCoding:
    def __init__(self,freq_dict: Dict[str,int]) -> None:
        self.freq_dict=freq_dict
        self.codes = {}
        self.root: Optional[HuffmanNode] = None
    
    def build_tree(self) -> None:
        heap: List[HuffmanNode] = [HuffmanNode(char, freq) for char, freq in self.freq_dict.items()]
        heapq.heapify(heap)
        while len(heap)>1:
            left = heapq.heappop(heap)
            right = heapq.heappop(heap)
            merged = HuffmanNode(None, left.freq + right.freq)
            merged.left = left
            merged.right = right
            heapq.heappush(heap, merged)
        self.root = heap[0]
    
    def assign_codes(self):
        self.codes.clear()
        def assign(node: Optional[HuffmanNode], code: str) -> None:
            if node is not None:
                if node.char is not None:
                    self.codes[node.char] = code
                    return
                assign(node.left, code + '0')
                assign(node.right, code + '1')
            else:
                return 
        assign(self.root,"")

    def encode(self, text: str) -> str:
        encoded = "".join(self.codes[char] for char in text)
        return encoded
    
    def decode(self, encoded_text: str) -> str:
        decoded = ""
        current_node = self.root
        for bit in encoded_text:
            if bit == '0':
                current_node = current_node.left
            else:
                current_node = current_node.right
            if current_node.char is not None:
                decoded += current_node.char
                current_node = self.root
        return decoded
    
def main():
    freq_dict = {'A':3,'B':4,'C':6,'D':2,'E':2}
    huffman_coding = HuffmanCoding(freq_dict)
    huffman_coding.build_tree()
    huffman_coding.assign_codes()
    print("Huffman Codes:", huffman_coding.codes)
    sample_text = "AABCCDEEDDEC"
    encoded_text = huffman_coding.encode(sample_text)
    decoded_text = huffman_coding.decode(encoded_text)
    print("Sample Text:", sample_text)
    print("Encoded Text:", encoded_text)
    print("Decoded Text:", decoded_text)

if __name__ == "__main__":
    main()